{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![title](img/genetics-bioinformatics.jpg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sequence Alignment with Dynamic Programming: Application of Smith–Waterman & Needleman–Wunsch algorithms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Section 1: Briefly Dynamic Programming\n",
    "\n",
    "\n",
    "Dynamic Programming is a programming approach used in order to increase efficiency of the solution process, by calculating the overlapping subproblems once and recording them for recurrent uses. There are two feasibility conditions that a problem needs to met before determining whether Dynamic Programming can be used.\n",
    "\n",
    "**These two conditions are:**\n",
    "\n",
    "1. Problem can be disintegrated into Overlapping Subproblems.\n",
    "2. Problem and its subproblems must show Optimal Substructure Property.\n",
    "\n",
    "## Overlapping Subproblems:\n",
    "\n",
    "If a problem can be dismantled in the subproblems that occur simultaneously or consecutively, it can be said that the problem has an Overlapping Subproblems property.\n",
    "\n",
    "## Optimal Substructure Property:\n",
    "\n",
    "If optimal solution to a problem can be achieved from the optimal solutions of its subproblems, it can be said that the problem has an Optimal Substructure property."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Section 2: Sequence Alignment Problem: Attempt to find the Ancestral Sequence\n",
    "\n",
    "Evolution has a huge impact on organisms and their current identity. By evolution, broadly speaking, nucleotide/protein compositions of organisms can change through mutation and selection over very long time. Mutations that cause these change can be classified in two types: substitution and insertion/deletion. \n",
    "\n",
    "## Substitution: \n",
    "\n",
    "As some sequences evolving through time, some components change to one another, this change referred as a substitution.\n",
    "\n",
    "## Insertion/Deletion: \n",
    "\n",
    "Through time, some other sequences might enter between the original sequence or some subsequences might disappear, this changes classified as insertion and deletion mutation.\n",
    "\n",
    "\n",
    "If a DNA sequence defined over an alphabet with letters {A,C,T,G}, these operations can be illustrated as a model below:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![title](img/SE-mutations.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, with these operations same genetical materials becomes slightly more different in closely related species. This difference expected to increase as relation become more ambiguous or distant. Scientist today can reach the genetic material of available species, but how about their ancestral genes? What is the original sequence? What operations possibly went underway, so that different sequences at different species occurred? What are their similarities and differences, and where they are exactly? These are the main questions in biology and sequence alignment tries to interpret these operations. \n",
    "\n",
    "But what is an alignment? Alignment of two sequence is simply a representation of these sequences put on top of each other, and pair them in a way that in the end both of them gets to same length. This is achieved by adding gaps to both sequences whenever felt necessary. For the general alignment cases, firstly global alignment must be taken into consideration. Briefly, global alignment tries to align elements of given sequences, specifically pair every element once among the sequences during the process. It is also called end-to-end alignment. Some feasibility conditions should be hold in order to call an alignment as global alignment. These feasibility constraints are:\n",
    "\n",
    "- **Sequences must be in same length in the end.**\n",
    "- **Dashes/gaps cannot be paired with itself between sequences.**\n",
    "- **If dashes are removed from sequences, same initial sequences are achieved.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![title](img/global-local-alignments.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now the global alignment problem and its feasibility conditions are known; but with these conditions, numerous global alignment can be achieved from given two sequences. That leads to next question: How an optimal alignment can be found? The alignment that might model ancestral sequence, or gives best insights about structural relations of given sequences. Hence an optimality condition must be set additional to the feasibility conditions. For optimality condition, setting a scoring scheme and evaluate the alignments according to this total score of the pairs is one solution. For the pairs that to be thought as a good match, positive score can be set at the matrix; and for the bad matches, or matches with \"dashes\", penalty score can be set. Thus created pairs can be scored and at the end total score of each alignment can be acquired. Alignment with the maximum score evaluated as optimal alignment among other alignments. Therefore optimality conditions can be created and manipulated whenever felt necessary.\n",
    "\n",
    "In Algorithmic thinking perspective, we can clarify the global sequence alignment problem a bit more by determining inputs and outputs of the problem. \n",
    "\n",
    "- **Our input:** Two sequences: X, Y and scoring matrix M.\n",
    "\n",
    "- **Our Output:** Global alignment sequences of X and Y and their maximum alignment score.\n",
    "\n",
    "Second alignment that will be used is local alignment. Some genes or their certain parts can be preserved better through time. In this case -where same genes of the related two species are taken into account- trying to find just these local parts and eliminating the rest is meaningful. Given two sequences, finding this partly alignment is called local alignment problem.\n",
    "\n",
    "More detailed explanations and mathematical representations of these alignments and their functional inputs and outputs can be found in here, provided by the course instructor."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Section 3: Algorithms Needleman–Wunsch and Smith–Waterman to solve global and local alignments:\n",
    "\n",
    "## Briefly the Algorithms:\n",
    "\n",
    "Needleman-Wunsch algorithm uses dynamic programming approach to tackle global alignment problem. It divides the problem to its subproblems, iterate and save the progress to an alignment matrix. Later it uses traceback method and finds the optimal global alignment.  \n",
    "\n",
    "Let's think of the following subproblem: \n",
    "\n",
    "**F(i, j)** represents the optimal alignment score between **X[1:i] and Y[1:j]**. **F(n, m)** will be optimal alignment score for global alignment problem, where **n, m** are the length of the sequences. And **F(i, j)** be can calculated directly if **F(i-1, j), F(i, j-1) and F(i-1,j-1)** known: \n",
    "\n",
    "**F(i, j) = max( F(i, j-1) + related penalty score, F(i-1, j)+related penalty score, F(i-1, j-1) + pair score).** \n",
    "We can think of this function as a **n x n matrix S**, where **S[i][j]** = **F(i, j)**. *In Needleman-Wunsch Algorithm*, after initializing the first row and column of **matrix S**, all other elements of the matrix calculated based on the formula above. Now the optimal global alignment score is stored in **S[n][m]**. In order to create the aligned sequences, traceback is initiated from the right bottom of the matrix.  For detailed information about the algorithm, please refer to this article.\n",
    "\n",
    "*Smith-Waterman algorithm* uses a variation of *Needleman-Wunsch algorithm*. In the alignment **matrix S**, it replaces negative values with zero and initiates the traceback not from the right bottom, but from the maximum entry. After traceback is initiated, it stops at the first zero entry it reaches. Since the local alignments have the maximum optimal alignment score,  it finds local alignments."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Implementation in Python:\n",
    "\n",
    "Function below takes both sequences, scoring matrix and global flag, which helps us to return local or global alignment matrix. According to global_flag, the function creates an alignment matrix S, in order to be used at the traceback phase."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_alignment_matrix(seq_x, seq_y, scoring_matrix, global_flag):\n",
    "    \"\"\"\n",
    "    Computes alligment matrix according to global flag, \n",
    "    with this matrix, max scores of possible subsequence pairs can be observed\n",
    "    \"\"\"\n",
    "    \n",
    "    lenght_x = len(seq_x)\n",
    "    lenght_y = len(seq_y)\n",
    "    alignment_matrix = [ [] for dummy_i in range(lenght_x+1)]\n",
    "    alignment_matrix[0].append(0)\n",
    "    \n",
    "    if global_flag == True: #If we will use for Needlwman-Wunsch: global alignment\n",
    "        for key_i in range(1,lenght_x+1): #column initialization\n",
    "            alignment_matrix[key_i].append(alignment_matrix[key_i-1][0] +  \\\n",
    "                                             scoring_matrix[seq_x[key_i-1]][\"-\"])\n",
    "        for key_j in range(1,lenght_y+1): #row initialization\n",
    "            alignment_matrix[0].append(alignment_matrix[0][key_j-1] + scoring_matrix[\"-\"][seq_y[key_j-1]])\n",
    "        \n",
    "        for key_i in range(1,lenght_x+1): #iteration\n",
    "            for key_j in range(1,lenght_y+1):\n",
    "                alignment_matrix[key_i].\\\n",
    "                    append(max(alignment_matrix[key_i-1][key_j-1] + scoring_matrix[seq_x[key_i-1]][seq_y[key_j-1]], \\\n",
    "                               alignment_matrix[key_i-1][key_j] + scoring_matrix[seq_x[key_i-1]][\"-\"], \\\n",
    "                               alignment_matrix[key_i][key_j-1] + scoring_matrix[\"-\"][seq_y[key_j-1]]))\n",
    "        return alignment_matrix\n",
    "    \n",
    "    else: # Case for Smith-Waterman: Local Alignments\n",
    "        for key_i in range(1,lenght_x+1):#column initialization assumend that gaps create negative values: set to 0\n",
    "            alignment_matrix[key_i].append(0) \n",
    "        for dummy_j in range(1,lenght_y+1): #row initialization assumend that gaps create negative values: set to 0\n",
    "            alignment_matrix[0].append(0 )\n",
    "        \n",
    "        for key_i in range(1,lenght_x+1): #iteration\n",
    "            for key_j in range(1,lenght_y+1):\n",
    "                alignment_matrix[key_i].\\\n",
    "                    append(max(0, #We dont want negative values\n",
    "                    alignment_matrix[key_i-1][key_j-1] + scoring_matrix[seq_x[key_i-1]][seq_y[key_j-1]], \\\n",
    "                    alignment_matrix[key_i-1][key_j] + scoring_matrix[seq_x[key_i-1]][\"-\"], \\\n",
    "                    alignment_matrix[key_i][key_j-1] + scoring_matrix[\"-\"][seq_y[key_j-1]]))\n",
    "        return alignment_matrix\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After the alignment matrix calculated, traceback can be done by the function below. This completes the Needleman-Wunsch algorithm, with this piece of code globally aligned pairs can be obtained."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_global_alignment(seq_x, seq_y, scoring_matrix, alignment_matrix):\n",
    "    \"\"\"\n",
    "    Completes the Needleman–Wunsch Algorithm, after alignment matrix phase completed.  \n",
    "    This function takes both sequences together with scoring matrix and alignment matrix\n",
    "    Starting from aligment matrix' last entry, it computes the sequuences according to 3 prior \n",
    "    possible starting points.\n",
    "    \"\"\"\n",
    "    lenght_x = len(seq_x)\n",
    "    lenght_y = len(seq_y)\n",
    "    x_fin = \"\" #start sequences\n",
    "    y_fin = \"\" #start sequences\n",
    "    final_score = alignment_matrix[lenght_x][lenght_y]\n",
    "    \n",
    "    while lenght_x <> 0 and lenght_y <> 0:\n",
    "        x_val, y_val = seq_x[lenght_x -1], seq_y[lenght_y -1]\n",
    "        poss_1 = alignment_matrix[lenght_x -1][lenght_y-1] + scoring_matrix[x_val][y_val] #first possiblity\n",
    "        \n",
    "        if alignment_matrix[lenght_x][lenght_y] == poss_1:\n",
    "            x_fin = x_val + x_fin\n",
    "            y_fin = y_val + y_fin\n",
    "            lenght_x -= 1\n",
    "            lenght_y -= 1\n",
    "        else:\n",
    "            poss_2 = alignment_matrix[lenght_x -1][lenght_y] + scoring_matrix[x_val][\"-\"]  #second possiblity  \n",
    "            \n",
    "            if alignment_matrix[lenght_x][lenght_y] == poss_2:\n",
    "                x_fin = x_val + x_fin\n",
    "                y_fin = \"-\" + y_fin\n",
    "                lenght_x -= 1\n",
    "            else:\n",
    "                x_fin = \"-\" + x_fin  \n",
    "                y_fin = y_val + y_fin\n",
    "                lenght_y -= 1\n",
    "    #if initial point could be reached, dashes will be added.            \n",
    "    while lenght_x <> 0:\n",
    "        x_val = seq_x[lenght_x -1]\n",
    "        x_fin = x_val + x_fin\n",
    "        y_fin = \"-\" + y_fin\n",
    "        lenght_x -= 1\n",
    "                \n",
    "    while lenght_y <> 0:\n",
    "        y_val = seq_y[lenght_y -1]\n",
    "        x_fin = \"-\" + x_fin  \n",
    "        y_fin = y_val + y_fin\n",
    "        lenght_y -= 1\n",
    "        \n",
    "    return (final_score, x_fin, y_fin)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For local alignments, traceback differs in terms of start and end points. Algorithm below, successfully calculates the local alignments and completes the Smith-Waterman algorithm.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_local_alignment(seq_x, seq_y, scoring_matrix, alignment_matrix): \n",
    "    \"\"\"\n",
    "    Completes the Smith–Waterman Algorithm, after alignment matrix phase completed. \n",
    "    \n",
    "    By setting global flag False at the allignment matrix and starting from the max entry,\n",
    "    this algorithm finds only local aligments.\n",
    "    \"\"\"\n",
    "    max_in_alligment = (0,0,0)\n",
    "    x_lenght = len(seq_x)\n",
    "    y_lenght = len(seq_y)\n",
    "    x_fin = \"\"\n",
    "    y_fin = \"\"\n",
    "    #Search for maximum\n",
    "    for key_i in range(1,x_lenght+1):\n",
    "        for key_j in range(1,y_lenght+1):\n",
    "            aligment_value = alignment_matrix[key_i][key_j]\n",
    "            if aligment_value > max_in_alligment[0]:\n",
    "                max_in_alligment = (aligment_value,key_i,key_j)\n",
    "                \n",
    "    score = max_in_alligment[0]\n",
    "    countdown_value =  max_in_alligment[0]\n",
    "    curr_pos_x, curr_pos_y = max_in_alligment[1], max_in_alligment[2]\n",
    "    \n",
    "    while countdown_value > 0: # while our value greater than 0, iterate.\n",
    "        x_val, y_val = seq_x[curr_pos_x-1], seq_y[curr_pos_y-1]\n",
    "        poss_1 = alignment_matrix[curr_pos_x -1][curr_pos_y-1] + scoring_matrix[x_val][y_val]\n",
    "        if alignment_matrix[curr_pos_x][curr_pos_y] == poss_1:\n",
    "            x_fin = x_val + x_fin\n",
    "            y_fin = y_val + y_fin\n",
    "            curr_pos_x -= 1\n",
    "            curr_pos_y -= 1\n",
    "            countdown_value = alignment_matrix[curr_pos_x][curr_pos_y]\n",
    "                \n",
    "        else:\n",
    "            poss_2 = alignment_matrix[curr_pos_x -1][curr_pos_y] + scoring_matrix[x_val][\"-\"]\n",
    "            if alignment_matrix[curr_pos_x][curr_pos_y] == poss_2:\n",
    "                x_fin = x_val + x_fin\n",
    "                y_fin = \"-\" + y_fin\n",
    "                curr_pos_x -= 1\n",
    "            else:\n",
    "                x_fin = \"-\" + x_fin  \n",
    "                y_fin = y_val + y_fin\n",
    "                curr_pos_y -= 1      \n",
    "    \n",
    "    return (score, x_fin, y_fin)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that the algorithms are ready, sequence alignment case can be introduced."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 4: Sequence Alignment Case and Helper Codes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Eyeless Gene, and it's correspondence in different species:\n",
    "\n",
    "\"In 1994, Walter Gehring and colleagues at the University of Basel carried out an \"interesting\" experiment: they were able to turn on a gene called eyeless in various places on the body of the fruit fly, Drosophila melanogaster. The result was astonishing - fruit flies developed that had whole eyes sprouting up all over their bodies. It turned out that the eyeless is a master regulatory gene - it controls a cascade that contains more than 2000 other genes. Turning it on anywhere in the body activates the cascade and produces a fully formed, but non-functioning, eye. Humans, as well as many other animals, have a slightly different version of the eyeless gene (that is, a similar, yet not identical sequence of the same gene). This observation suggests that about 600 million years ago (the estimated time of divergence between humans and fruit flies) there was an ancestral organism that itself used some version of eyeless, and that throughout the evolution of humans and fruit flies this gene continued to be maintained, albeit while accumulating mutations that did not greatly affect its function. In particular, a substring of the eyeless protein of about 130 amino acids, known as the PAX domain, whose function is to bind specific sequences of DNA, is virtually identical between the human and fruit fly versions of eyeless.\"*\n",
    "\n",
    "Genes and scoring matrix are given as URL below, and possible alignments will be scored according to this given PAM50 matrix. Helper functions below will be used to read proteins and PAM50 scoring matrix. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Genes and scoring matrix are given as URL below, and possible alignments will be scored according to this given PAM50 matrix. Helper functions below will be used to read proteins and PAM50 scoring matrix.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![title](img/scoring-matrix.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Helper Functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "This code is provided by instructors from Rice University\n",
    "by online Algorithmic Thinking course through online learning platform Coursera.\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "import math\n",
    "import random\n",
    "import urllib2\n",
    "\n",
    "\n",
    "\n",
    "# URLs for data files\n",
    "PAM50_URL = \"http://storage.googleapis.com/codeskulptor-alg/alg_PAM50.txt\"\n",
    "HUMAN_EYELESS_URL = \"http://storage.googleapis.com/codeskulptor-alg/alg_HumanEyelessProtein.txt\"\n",
    "FRUITFLY_EYELESS_URL = \"http://storage.googleapis.com/codeskulptor-alg/alg_FruitflyEyelessProtein.txt\"\n",
    "CONSENSUS_PAX_URL = \"http://storage.googleapis.com/codeskulptor-alg/alg_ConsensusPAXDomain.txt\"\n",
    "WORD_LIST_URL = \"http://storage.googleapis.com/codeskulptor-assets/assets_scrabble_words3.txt\"\n",
    "\n",
    "\n",
    "\n",
    "###############################################\n",
    "# provided code\n",
    "\n",
    "def read_scoring_matrix(filename):\n",
    "    \"\"\"\n",
    "    Read a scoring matrix from the file named filename.  \n",
    "\n",
    "    Argument:\n",
    "    filename -- name of file containing a scoring matrix\n",
    "\n",
    "    Returns:\n",
    "    A dictionary of dictionaries mapping X and Y characters to scores\n",
    "    \"\"\"\n",
    "    scoring_dict = {}\n",
    "    scoring_file = urllib2.urlopen(filename)\n",
    "    ykeys = scoring_file.readline()\n",
    "    ykeychars = ykeys.split()\n",
    "    for line in scoring_file.readlines():\n",
    "        vals = line.split()\n",
    "        xkey = vals.pop(0)\n",
    "        scoring_dict[xkey] = {}\n",
    "        for ykey, val in zip(ykeychars, vals):\n",
    "            scoring_dict[xkey][ykey] = int(val)\n",
    "    return scoring_dict\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def read_protein(filename):\n",
    "    \"\"\"\n",
    "    Read a protein sequence from the file named filename.\n",
    "\n",
    "    Arguments:\n",
    "    filename -- name of file containing a protein sequence\n",
    "\n",
    "    Returns:\n",
    "    A string representing the protein\n",
    "    \"\"\"\n",
    "    protein_file = urllib2.urlopen(filename)\n",
    "    protein_seq = protein_file.read()\n",
    "    protein_seq = protein_seq.rstrip()\n",
    "    return protein_seq\n",
    "\n",
    "\n",
    "def read_words(filename):\n",
    "    \"\"\"\n",
    "    Load word list from the file named filename.\n",
    "\n",
    "    Returns a list of strings.\n",
    "    \"\"\"\n",
    "    # load assets\n",
    "    word_file = urllib2.urlopen(filename)\n",
    "    \n",
    "    # read in files as string\n",
    "    words = word_file.read()\n",
    "    \n",
    "    # template lines and solution lines list of line string\n",
    "    word_list = words.split('\\n')\n",
    "    print \"Loaded a dictionary with\", len(word_list), \"words\"\n",
    "    return word_list\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Section 5: Analysis of the case\n",
    "\n",
    "**The task is to determine local or global similarities between human and fruit fly eyeless gene sequences. Compare them by Consensus PAX gene, calculate their agreement and do hypothesis testing over the results to determine the likelihood of randomness. For finding global alignments, *Needleman–Wunsch algorithm* and for the local alignments *Smith–Waterman algorithm* will be used.**\n",
    "\n",
    "## Comparison of Human and Fruit Fly Eyeless Genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Human Eyeless Gene:  MQNSHSGVNQLGGVFVNGRPLPDSTRQKIVELAHSGARPCDISRILQVSNGCVSKILGRYYETGSIRPRAIGGSKPRVATPEVVSKIAQYKRECPSIFAWEIRDRLLSEGVCTNDNIPSVSSINRVLRNLASEKQQMGADGMYDKLRMLNGQTGSWGTRPGWYPGTSVPGQPTQDGCQQQEGGGENTNSISSNGEDSDEAQMRLQLKRKLQRNRTSFTQEQIEALEKEFERTHYPDVFARERLAAKIDLPEARIQVWFSNRRAKWRREEKLRNQRRQASNTPSHIPISSSFSTSVYQPIPQPTTPVSSFTSGSMLGRTDTALTNTYSALPPMPSFTMANNLPMQPPVPSQTSSYSCMLPTSPSVNGRSYDTYTPPHMQTHMNSQPMGTSGTTSTGLISPGVSVPVQVPGSEPDMSQYWPRLQ\n",
      "\n",
      "Fruitfly Eyeless Gene:  MRNLPCLGTAGGSGLGGIAGKPSPTMEAVEASTASHPHSTSSYFATTYYHLTDDECHSGVNQLGGVFVGGRPLPDSTRQKIVELAHSGARPCDISRILQVSNGCVSKILGRYYETGSIRPRAIGGSKPRVATAEVVSKISQYKRECPSIFAWEIRDRLLQENVCTNDNIPSVSSINRVLRNLAAQKEQQSTGSGSSSTSAGNSISAKVSVSIGGNVSNVASGSRGTLSSSTDLMQTATPLNSSESGGASNSGEGSEQEAIYEKLRLLNTQHAAGPGPLEPARAAPLVGQSPNHLGTRSSHPQLVHGNHQALQQHQQQSWPPRHYSGSWYPTSLSEIPISSAPNIASVTAYASGPSLAHSLSPPNDIESLASIGHQRNCPVATEDIHLKKELDGHQSDETGSGEGENSNGGASNIGNTEDDQARLILKRKLQRNRTSFTNDQIDSLEKEFERTHYPDVFARERLAGKIGLPEARIQVWFSNRRAKWRREEKLRNQRRTPNSTGASATSSSTSATASLTDSPNSLSACSSLLSGSAGGPSVSTINGLSSPSTLSTNVNAPTLGAGIDSSESPTPIPHIRPSCTSDNDNGRQSEDCRRVCSPCPLGVGGHQNTHHIQSNGHAQGHALVPAISPRLNFNSGSFGAMYSNMHHTALSMSDSYGAVTPIPSFNHSAVGPLAPPSPIPQQGDLTPSSLYPCHMTLRPPPMAPAHHHIVPGDGGRPAGVGLGSGQSANLGASCSGSGYEVLSAYALPPPPMASSSAADSSFSAASSASANVTPHHTIAQESCPSPCSSASHFGVAHSSGFSSDPISPAVSSYAHMSYNYASSANTMTPSSASGTSAHVAPGKQQFFASCFYSPWV\n"
     ]
    }
   ],
   "source": [
    "human_gene = read_protein(HUMAN_EYELESS_URL)\n",
    "fruitfly_gene = read_protein(FRUITFLY_EYELESS_URL)\n",
    "pam_50_matrix = read_scoring_matrix(PAM50_URL)\n",
    "\n",
    "print \"Human Eyeless Gene: \", human_gene\n",
    "print \n",
    "print \"Fruitfly Eyeless Gene: \", fruitfly_gene\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As can be seen, length of both sequences differ significantly. It would be reasonable to search for local similarities using local alignment function, rather them align them end-to-end with global alignment function."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Local Alignment between sample genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "875\n",
      "HSGVNQLGGVFVNGRPLPDSTRQKIVELAHSGARPCDISRILQVSNGCVSKILGRYYETGSIRPRAIGGSKPRVATPEVVSKIAQYKRECPSIFAWEIRDRLLSEGVCTNDNIPSVSSINRVLRNLASEK-QQ\n",
      "HSGVNQLGGVFVGGRPLPDSTRQKIVELAHSGARPCDISRILQVSNGCVSKILGRYYETGSIRPRAIGGSKPRVATAEVVSKISQYKRECPSIFAWEIRDRLLQENVCTNDNIPSVSSINRVLRNLAAQKEQQ\n"
     ]
    }
   ],
   "source": [
    "alignment_matrix = compute_alignment_matrix(human_gene, fruitfly_gene, pam_50_matrix, False)\n",
    "alignment_loc = compute_local_alignment(human_gene, fruitfly_gene, pam_50_matrix, alignment_matrix)\n",
    "\n",
    "for element in alignment_loc:\n",
    "    print element"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Local alignment function determine significance similarities between human and fruit fly eyeless genes with the score of 875. When both local sequences examined, only small differences noticed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison to Consensus PAX Domain:\n",
    "\n",
    "After local similar parts are found, comparison of these gene parts with consensus gene -gene that corresponds the Eyeless PAX6 domain of different species- would also answer the question of their similarity to other PAX6 domains. Consensus domain:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GHGGVNQLGGVFVNGRPLPDVVRQRIVELAHQGVRPCDISRQLRVSHGCVSKILGRYYETGSIKPGVIGGSKPKVATPKVVEKIAEYKRQNPTMFAWEIRDRLLAERVCDNDTVPSVSSINRIIR\n"
     ]
    }
   ],
   "source": [
    "CONSENSUS_PROTEIN = read_protein(CONSENSUS_PAX_URL)\n",
    "print CONSENSUS_PROTEIN"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For these purpose, functions below are created. First, dashes should be removed, then results of each sequence should be globally aligned with consensus Pax. After alignment created, their agreement should be calculated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def delete_dashes(string):\n",
    "    \"\"\"\n",
    "    delete dashes from given string\n",
    "    \"\"\"\n",
    "    dashless_str = \"\"\n",
    "    for char in string:\n",
    "        if char == \"-\":\n",
    "            continue\n",
    "        else:\n",
    "            dashless_str += char  \n",
    "    return dashless_str\n",
    "\n",
    "def calculate_seqs_agreement(seq1,seq2):\n",
    "    \"\"\"\n",
    "    Given two same lenght sequences, this function calculates the perccentage of \n",
    "    the agreeing elements over the whole sequences\n",
    "    \"\"\"\n",
    "    assert len(seq1) == len(seq2)\n",
    "    seq_lenght = len(seq1)\n",
    "    agr_number = 0\n",
    "    \n",
    "    for key_i in range(seq_lenght):\n",
    "        if seq1[key_i] == seq2[key_i]:\n",
    "            agr_number += 1\n",
    "    \n",
    "    percentage = agr_number/float(seq_lenght)*100   \n",
    "    return percentage\n",
    "\n",
    "def compare_seq_cons(seq1, consensus):\n",
    "    \"\"\"\n",
    "    Function that takes sequence and consensus sequence\n",
    "    Remove their dashes, align them globally and measure the similarity percentage.\n",
    "    Returns their similarity percentage\n",
    "    \"\"\"\n",
    "    seq1_dashless = delete_dashes(seq1)\n",
    "    align_m = compute_alignment_matrix(seq1_dashless, consensus, pam_50_matrix, True)\n",
    "    alignments= compute_global_alignment(seq1_dashless, consensus, pam_50_matrix, align_m)\n",
    "    percentage = calculate_seqs_agreement(alignments[1], alignments[2])\n",
    "    \n",
    "    return percentage\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Human and Fly alignment sequences recorded previously on variable alignment_loc. Their agreement with consensus domain:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Human-Consensus Agreement Percentage:  72.9323308271\n",
      "Fly-Consensus Agreement Percentage:  70.1492537313\n"
     ]
    }
   ],
   "source": [
    "human_cons_perc = compare_seq_cons(alignment_loc[1],CONSENSUS_PROTEIN)\n",
    "fly_cons_perc = compare_seq_cons(alignment_loc[2],CONSENSUS_PROTEIN)\n",
    "print \"Human-Consensus Agreement Percentage: \", human_cons_perc\n",
    "print \"Fly-Consensus Agreement Percentage: \", fly_cons_perc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As can be seen similarity is around **70-72%**. Length of the consensus gene is **125**, and to get mentioned percentage, at least **88** amino acid agreements needed. Also if the event type is considered, two sequences with an at least **%70** agreement refers a to binomial process with the success rate **1/24** (23 amino acids and a gap). If its calculated, the result would be **10^(-91)**.  As can be seen from the result, the odds are extremely unlikely."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hypothesis Testing\n",
    "\n",
    "At the last calculation, appearance chance of every protein in the sequence was assumed to be equal. But this is not the case: to avoid this situation and create a better model, a simulation would be more suitable. At this simulation, fruit fly gene will be shuffled while human genome remains the same at every iteration. At each iteration, human and shuffled fruit fly genome will be globally aligned, and their score will be recorded. After repetition of 1000 experiments, scores will be recorded according to their appearance fraction, and later the distribution will be normalized. This function can be seen as:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_null_distribution(seq_x, seq_y, scoring_matrix, num_trials):\n",
    "    \"\"\"\n",
    "    using the sequences, holds trials by randomize one of the given sequence \n",
    "    and compute newly created pairs score. Returns the unnormalized distribution of these scores. \n",
    "    \"\"\"\n",
    "    result_dict = {}\n",
    "    \n",
    "    for trial_num in range(num_trials):\n",
    "        seq_y_copy = (seq_y+\".\")[:-1]\n",
    "        \n",
    "        seq_y_copy = list(seq_y_copy)\n",
    "        random.shuffle(seq_y_copy)\n",
    "        \"\".join(seq_y_copy)\n",
    "        \n",
    "        align_m = compute_alignment_matrix(seq_x, seq_y_copy, scoring_matrix, False)\n",
    "        trail_result = compute_local_alignment(seq_x,seq_y_copy,scoring_matrix, align_m)\n",
    "        if trail_result[0] in result_dict.keys():\n",
    "            result_dict[trail_result[0]] += 1\n",
    "        else:\n",
    "            result_dict[trail_result[0]] = 1\n",
    "           \n",
    "    return result_dict    \n",
    "\n",
    "def normalize_distribution(dist):\n",
    "    \"\"\"\n",
    "    normalize the entered distribution given as dictionary \n",
    "    \"\"\"\n",
    "    total = 0\n",
    "    dist_copy = dist.copy()\n",
    "    \n",
    "    for key in dist:\n",
    "        total += dist[key]\n",
    "    \n",
    "    for key in dist:\n",
    "        dist_copy[key] = dist[key]/float(total)\n",
    "        \n",
    "    return dist_copy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If matplotlib is used to create a bar chart to examine the distribution. Following code can be created."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA50AAAFdCAYAAACAWwnXAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3Xm8JFV9///XW4bFDVAcF2BYFFAhcR1Ro9EoiqACxmCEuOCKmhC3xIjxG0USo6gR/bmEoKCIRkDUOG4BI65RkQFFRcSMCDLiMuybKIOf3x91LjaX7jvNzK3bd+68no9HP25X1amqT1Wf7tufPqdOpaqQJEmSJKkPt5l0AJIkSZKkhcukU5IkSZLUG5NOSZIkSVJvTDolSZIkSb0x6ZQkSZIk9cakU5IkSZLUG5NOSVoPJKkkO006jj4l+bMkKwemz0nyZ7O8jw8m+Zcxy+7QzvuiNv35JAfNZjxtu7dN8ukkVyb52Gxvf8j+jkryT33vZ1xJ7p3kO0muTvLSMddZcO+H+fa6SNJsWjTpACRpEpI8EngLsBtwI3Au8PKqOmOigekmVbXbpGMYVFV7j1MuSQE7V9WKMTe9P3A3YKuqWr228Y2I5TnAC6rqkVPzqurFs7mPWfAPwJer6oHDFib5MvDhqnr/nEY1xyb1uiQ5DNipqp45if1L2jDY0ilpg5Nkc+AzwLuAOwPbAG8AfjvL+9loNrc330y1AGqdbQ/8eFTCuQGc5+2BcyYdxCQt9M8KSTLplLQh2gWgqj5aVTdW1W+q6tSq+t5UgSQvTHJu6/L3wyQPavPvm+TLSa5o3T/3HVjng0n+PcnnklwLPCbJpkneluRnSX7VutDdtpW/S5LPtG1dluRrSWb6XH5ikvOTXJLkrUlu07Z/WZI/Hojjrkl+k2Tx9A0keU6Sr7eYLk/y0yR7DyzfOsmyts0VSV44sOywJCcn+XCSq4DntHkfa/OuTvL9JLskeU2SXye5KMmeA9t47sB5PT/Ji0YdbJILkjyuPb8iyTXtcW3rXrlDW/bkJN9tZb6R5H4D23hgkrPa/k4ENpthfxu183JJkvOBJ01b/uUkL2jPd0ryldYl9pK2bZJ8tRU/u8X69FH7a+XfALwOeHor//z2Gv1vkiOTXAYcluReSU5Lcmnb30eSbDmwnSVJPpFkVSvz7iT3BY4CHt62fUUre7Muxq2ur2iv+bIkWw8sqyQvTvJ/rb68J0lmOgcjjnPfdO+XK9p5vG+bfxrwGODdLcZdpq33RuBPB5a/e2Dx44bF1dZ7Xqtnlyc5Jcn2M8T2sFZvrkhydlqX7iR3TrIyyT5t+g7tPD174DweleQLrX59ZXA/Se7Tll2W5LwkfzmwbNhnxU2vS1pX8yT/kO599IskT0nyxCQ/btv8x4Ht3SbJoUl+0l7/k5LcuS2b6iZ+ULrPoUuSvLYt2wv4R/5Q/84edZ4kaZ1UlQ8fPnxsUA9gc+BS4Dhgb+BO05Y/Dfg58BAgwE50rTEbAyvovqRtAjwWuBq4d1vvg8CVwCPoftTbDHgHsIyuRfWOwKeBN7Xyb6JLCjZujz8FMiLmAr7UtrMd8GO6bpMA7wWOGCj7MuDTI7bzHOAG4IXARsBLgIun9gt8pW1vM+ABwCpgj7bssLbuU9rx3bbNux54At0lGx8Cfgq8th3TC4GfDuz/ScC92nl9NHAd8KC27M+AlQNlLwAeN+QY/hX4atv+g4BfAw9tx3NQW2/T9hpdCLyild2/xf8vI87Ni4EfAUvaef5SO++L2vIvD5zzj7ZjnHqdHznttdrpVtTHw+i6jw6+RquBv23n9LZ0dfDx7bgWt+N/Ryu/EXA2cCRw+8F42ra+Pm1/H5w6B3R1+JJ2Hjela/3/6rRj+QywJV29WwXstaZzMG1/uwDXtvg3putOuwLYZPp5HbH+LZavIa6ntO3ft52//wd8Y8S2t6H7LHhiO47Ht+nFbfmewC+BuwLvA06edh6vBh7Vzt07p851ex0uAp7bYnhQO8+7zfBZMfi6/FmrA6/jD++jVcB/0n2O7Eb3vrtnK/9y4FvAti2W/wA+2pbt0M7X++jq0v3penXcd1j98+HDh48+HrZ0StrgVNVVwCP5wxexVa2F526tyAuAt1TVGdVZUVUXAg8D7gC8uap+V1Wn0X3xPXBg85+qqv+tqt/TfbF7IfCKqrqsqq6mS5gOaGVvAO4BbF9VN1TV16qqZgj9iLadn9Els1P7PQ74q/yhlfRZwPEzbOfCqnpfVd3Y1r0HcLckS9p5eXVVXV9V3wXe37Y35ZtV9V9V9fuq+k2b97WqOqW67qEfo0uK3lxVNwAnADtMtcpV1Wer6iftvH4FOJUu2R5Lazn8K+Av2vZfCPxHVZ1eXav1cXTn/WHtsTFdcnZDVZ0MzHTN7l+2shdV1WV0PwqMcgPdDxFbt3P19XGPYUwXV9W7qmp1dS3xK6rqC1X126paBbydLmkH2B3YGnhVVV17K+N5BnBsVZ1VVb8FXkPXMrrDQJk3V9UVrd59ie7HCBj/HDwd+GyL/wbgbXTJz5+MGeMoo+J6Ed0PO+e2OvmvwANGtHY+E/hcVX2u1ekvAMvpklCq6lS6Ov1Fuh9MprfMf7aqvtrO3Wvpzt0S4MnABVX1gfYangV8nO6Hjyk3fVZU1fVDYrsBeOPA++guwDur6uqqOoeuS/JUq/6LgNdW1coWy2HA/rl51+w3tLp0Nt2PFPcfcV4ladaZdEraILUvpM+pqm2BP6L70v6OtngJ8JMhq20NXNQSyikX0rWWTLlo4Pli4HbAma3r3hXAf7f5AG+la5E5NV1X00PXEPbgti9s8VBVp9O1JD06yX3oWsWWzbCdX049qarr2tM7tO1NJcfjHN+UXw08/w1wSUtop6antk+SvZN8q3UPvILuy/1dZoj1JkkeCLwb+POWeEGX9Pzd1Plt21zSjmVr4OfTEvkLZ9jF1tzyHI/yD3Sttd9u3UafN84x3Ao3O8/pukyfkOTn6bo2f5g/nLcldD8krM0gRFszcJxVdQ1dS9/ga/7LgefX0V5Lxj8H0/fxe7rj22ZE+XGNimt74J0D9eGyFuew/W0PPG1a/Xkk3Q8xU46m+4z4QFVdOm39m16ndu4uozve7YGHTtvuM4C7D1t3hEuHvI+mv9cGj/mTA/s6l26AtLsNlB91viSpdyadkjZ4VfUjuq5tf9RmXUTXBXS6i4Elufl1l9vRdcW9aXMDzy+h+2K4W1Vt2R5bVNUd2n6vrqq/q6p7AvsAr0yyxwyhLpm234sHpo+ja7V5Fl0XwGEtJ2tyMXDnJHectp9Rx3erJNmUrrXnbcDdqmpL4HN0CcGa1l0MfBI4pKq+M7DoIrrWoC0HHrerqo8CvwC2GbzWrx3PKL/glud4qKr6ZVW9sKq2pmtlem9m9xYe08/zm9q8+1XV5nSv9dRxXQRsl+EDDq3p9bqYLmEBIMntga24+Ws+PMDxz8H0fYTuPK9xH1O7GrPclIuAF02rE7etqm+MKHv8tLK3r6o3t1g3ouuq+iHgJUOO76b6kuQOdN2yL27b/cq07d6hql6yDsc1k4uAvaftb7OqGuccz2YckjSUSaekDU4b4OPvkmzbppfQdVX9VivyfuDvkzw4nZ1a17ypFsV/SLJxG3BkH7qub7fQWnTeBxyZ5K5tX9skeUJ7/uS27QBX0bVM3DhsW82rktypxfsyYHDgluOBP6dLRj50a89Ji/ci4BvAm5Jslm5AnucDH1mb7Q2xCd31ZquA1ekGMNpz5lVuGr3148BHqmr6YDXvA16c5KHttbp9kie1xPmbdNfFvTTJoiRPpeuKOspJrey2Se4EjGx5TvK0qfoDXE73xX3qtfsVcM81HdetdEfgGuCKJNsArxpY9m26hPnN7fg3S/KIgVi2TbLJiO3+J/DcJA9oPwr8K3B6VV2wpoDWcA4GnQQ8KckeSTYG/o6uC/SwJHCYW3s+jwJek2S3FucWSZ42ouyHgX2SPCHdQFKbpRvEZ+q4pgbreR7djyUfys1Hmn1ikke28/vPdOfuIrpu97skeVb7rNg4yUPSBlDqwVHAG6e6ECdZnGS/Mdf9FV0XeL8TSuqNHzCSNkRX0w08c3q6kSO/BfyA7sswVfUx4I10X8ivBv4LuHNV/Q7Yl27woUvoBtx5dmspHeXVdF1ov9W6Rf4PcO+2bOc2fQ1dgvTeqvryDNv6FHAm8F3gs8AxUwuqaiVwFt0X/6+NcxJGOJBu4JGL6VoWX9+uc1tnrdvuS+mSkMvprs2cqRvwlG3prvt8ef4wgu01SbarquV013W+u21zBd3gObTX66lt+nK6aws/McN+3gecQne921lrKPsQuvpzTTuGl1XVT9uyw4DjWlfHv0yy3VS8YxzrKG+gG4zmSrrX/qbYWhfMfei6Vf8MWEl3rACn0V3798skl0zfaFV9EfgnuqT+F3Qt/AdMLzfCTOdgcB/n0f0Y8i66980+wD7t9RnHO+muT7w8yf+3psJV9UngCOCE9p77Ad17dljZi4D96JLLVXQthq8CbpPkwcAr6d7jN7ZtFjf/MeI/gdfTdat9MF0X2qm6vifdubyYrmvrEXQ/uvThnXSvwalJrqb7THvomOt+rP29NMlZfQQnSVOjFUqS1nNJjqUbgOb/TToWaaFL8kG60ZZ9v0nSGiz0G05L0gYh3WijTwUeONlIJEmSbs7utZK0nkvyz3RdCN86rHujJEnSJNm9VpIkSZLUG1s6JUmSJEm9MemUJEmSJPVmwQwkdJe73KV22GGHSYchSZIkSQvemWeeeUlVLR6n7IJJOnfYYQeWL18+6TAkSZIkacFLcuG4Ze1eK0mSJEnqjUmnJEmSJKk3Jp2SJEmSpN6YdEqSJEmSemPSKUmSJEnqjUmnJEmSJKk3Jp2SJEmSpN6YdEqSJEmSemPSKUmSJEnqjUmnJEmSJKk3Jp2SJEmSpN6YdEqSJEmSerNo0gFIuqVk7darmt04JEmSpHVlS6ckSZIkqTcmnZIkSZKk3vSadCbZK8l5SVYkOXTI8k2TnNiWn55khzZ/4yTHJfl+knOTvKbPOCVJkiRJ/egt6UyyEfAeYG9gV+DAJLtOK/Z84PKq2gk4EjiizX8asGlV/THwYOBFUwmpJEmSJGn90WdL5+7Aiqo6v6p+B5wA7DetzH7Ace35ycAeSQIUcPski4DbAr8DruoxVkmSJElSD/pMOrcBLhqYXtnmDS1TVauBK4Gt6BLQa4FfAD8D3lZVl03fQZKDkyxPsnzVqlWzfwSSJEmSpHXSZ9I57KYP02/oMKrM7sCNwNbAjsDfJbnnLQpWHV1VS6tq6eLFi9c1XkmSJEnSLOsz6VwJLBmY3ha4eFSZ1pV2C+Ay4K+A/66qG6rq18D/Akt7jFWSJEmS1IM+k84zgJ2T7JhkE+AAYNm0MsuAg9rz/YHTqqroutQ+Np3bAw8DftRjrJIkSZKkHvSWdLZrNA8BTgHOBU6qqnOSHJ5k31bsGGCrJCuAVwJTt1V5D3AH4Ad0yesHqup7fcUqSZIkSepHuobF9d/SpUtr+fLlkw5DmhUZdrXzGBbI21mSJEnzXJIzq2qsSyD77F4rSZIkSdrAmXRKkiRJknpj0ilJkiRJ6o1JpyRJkiSpNyadkiRJkqTemHRKkiRJknpj0ilJkiRJ6o1JpyRJkiSpNyadkiRJkqTemHRKkiRJknpj0ilJkiRJ6s2iSQcgaX5L1m69qtmNQ5IkSesnWzolSZIkSb0x6ZQkSZIk9cakU5IkSZLUG5NOSZIkSVJvTDolSZIkSb0x6ZQkSZIk9cakU5IkSZLUG5NOSZIkSVJvTDolSZIkSb0x6ZQkSZIk9cakU5IkSZLUG5NOSZIkSVJvek06k+yV5LwkK5IcOmT5pklObMtPT7JDm/+MJN8dePw+yQP6jFWSJEmSNPt6SzqTbAS8B9gb2BU4MMmu04o9H7i8qnYCjgSOAKiqj1TVA6rqAcCzgAuq6rt9xSpJkiRJ6kefLZ27Ayuq6vyq+h1wArDftDL7Ace15ycDeyTJtDIHAh/tMU5JkiRJUk/6TDq3AS4amF7Z5g0tU1WrgSuBraaVeTomnZIkSZK0Xuoz6ZzeYglQt6ZMkocC11XVD4buIDk4yfIky1etWrX2kUqSJEmSetFn0rkSWDIwvS1w8agySRYBWwCXDSw/gBlaOavq6KpaWlVLFy9ePCtBS5IkSZJmT59J5xnAzkl2TLIJXQK5bFqZZcBB7fn+wGlVVQBJbgM8je5aUEmSJEnSemhRXxuuqtVJDgFOATYCjq2qc5IcDiyvqmXAMcDxSVbQtXAeMLCJRwErq+r8vmKUFrJbDMk1ppreCV6SJElaB6kF8g1z6dKltXz58kmHIc2K2UgYZyvpNHmVJEnSdEnOrKql45Tts3utJEmSJGkDZ9IpSZIkSeqNSackSZIkqTcmnZIkSZKk3ph0SpIkSZJ609stUyRpIXNUX0mSpPHY0ilJkiRJ6o1JpyRJkiSpNyadkiRJkqTemHRKkiRJknpj0ilJkiRJ6o1JpyRJkiSpNyadkiRJkqTemHRKkiRJknpj0ilJkiRJ6s2iSQcgSeNK1m69qtmNQ5IkSeOzpVOSJEmS1BuTTkmSJElSb0w6JUmSJEm9MemUJEmSJPXGpFOSJEmS1Js1Jp1J3pJk8yQbJ/likkuSPHMugpOkhS5Zu4ckSdL6YpyWzj2r6irgycBKYBfgVb1GJUmSJElaEMZJOjduf58IfLSqLusxHkmSJEnSArJojDKfTvIj4DfAXydZDFzfb1iSJEmSpIVgjS2dVXUo8HBgaVXdAFwL7DfOxpPsleS8JCuSHDpk+aZJTmzLT0+yw8Cy+yX5ZpJzknw/yWbjHpQkSZIkaX4Y2dKZ5KlD5g1OfmKmDSfZCHgP8Hi6a0HPSLKsqn44UOz5wOVVtVOSA4AjgKcnWQR8GHhWVZ2dZCvghjGPSZIkSZI0T8zUvXafGZYVa0g6gd2BFVV1PkCSE+haSAeTzv2Aw9rzk4F3p8ts9wS+V1VnA1TVpWvYlyRJkiRpHhqZdFbVc9dx29sAFw1MrwQeOqpMVa1OciWwFd0IuZXkFGAxcEJVvWX6DpIcDBwMsN12261juJIkSZKk2TbOQEIkeRKwG3DTdZVVdfiaVhsyr8Ysswh4JPAQ4Drgi0nOrKov3qxg1dHA0QBLly6dvm1JkiRJ0oStcSChJEcBTwf+li5JfBqw/RjbXgksGZjeFrh4VJl2HecWwGVt/leq6pKqug74HPCgMfYpSZIkSZpHxrlP559U1bPpBvx5A91ItkvWsA7AGcDOSXZMsglwALBsWpllwEHt+f7AaVVVwCnA/ZLcriWjj+bm14JKkiRJktYD43Sv/U37e12SrYFLgR3XtFK7RvMQugRyI+DYqjonyeHA8qpaBhwDHJ9kBV0L5wFt3cuTvJ0ucS3gc1X12Vt5bJIkSZKkCRsn6fxMki2BtwJn0SWB7x9n41X1ObqusYPzXjfw/Hq67rrD1v0w3W1TJEmSJEnrqTUmnVX1z+3px5N8Btisqq7sNyxJkiRJ0kIwMulM8tiqOi3JU4cso6rWdJ9OSZIkSdIGbqaWzkcDpwH7DFlWgEmnJEmSJGlGI5POqnp9ktsAn6+qk+YwJkmSJEnSAjHjLVOq6vfAIXMUiyRJkiRpgRnnPp1fSPL3SZYkufPUo/fIJEmSJEnrvXFumfK89vdvBuYVcM/ZD0eSJEmStJCMk3Tet91P8yZJNuspHkmSJEnSAjJO99pvjDlPkiRJkqSbmek+nXcHtgFum+SBQNqizYHbzUFskiRJkqT13Ezda58APAfYFvg3/pB0XgX8Y79hSZIkSZIWgpnu03kccFySv6iqj89hTJIkSZKkBWKN13SacEqSJEmS1tY4AwlJkiRJkrRWTDolSZIkSb2ZafTap860YlV9YvbDkSRJkiQtJDONXrvPDMsKMOmUJEmSJM1optFrnzuXgUiSJEmSFp6ZWjpvkuRJwG7AZlPzqurwvoKSJEmSJC0MaxxIKMlRwNOBvwUCPA3Yvue4JEmSJEkLwDij1/5JVT0buLyq3gA8HFjSb1iSJEmSpIVgnO61v2l/r0uyNXApsGN/IUlSv5K1W69qduOYLQvteCRJ0sIyTtL5mSRbAm8FzqIbufb9vUYlSZIkSVoQxkk631JVvwU+nuQzdIMJXd9vWJIkSZKkhWCcazq/OfWkqn5bVVcOzptJkr2SnJdkRZJDhyzfNMmJbfnpSXZo83dI8psk322Po8Y7HEmSJEnSfDKypTPJ3YFtgNsmeSDdyLUAmwO3W9OGk2wEvAd4PLASOCPJsqr64UCx59MNULRTkgOAI+hGygX4SVU94NYekCRJkiRp/pipe+0TgOcA2wJvH5h/NfCPY2x7d2BFVZ0PkOQEYD9gMOncDzisPT8ZeHeytkNiSJIkSZLmm5FJZ1UdBxyX5C+q6uNrse1tgIsGplcCDx1VpqpWJ7kS2Kot2zHJd4CrgP9XVV9bixgkSZIkSRM0zkBCX0zyduBRbforwOHt2s6ZDGuxnD5A/6gyvwC2q6pLkzwY+K8ku1XVVTdbOTkYOBhgu+22W0M4kiRJkqS5Ns5AQsfQdan9y/a4CvjAGOutBJYMTG8LXDyqTJJFwBbAZW3AoksBqupM4CfALtN3UFVHV9XSqlq6ePHiMUKSJEmSJM2lcZLOe1XV66vq/PZ4A3DPMdY7A9g5yY5JNgEOAJZNK7MMOKg93x84raoqyeI2EBFJ7gnsDJw/zgFJkiRJkuaPcbrX/ibJI6vq6wBJHgH8Zk0rtWs0DwFOATYCjq2qc5IcDiyvqmV0rajHJ1kBXEaXmELXlffwJKuBG4EXV9Vlt/bgJEmSJEmTlarpl1lOK5DcH/gQXddXgMuBg6rqez3HdqssXbq0li9fPukwpFmxtmM4D76dZ2Mbs7md2TCfjmkhxiJJkjSuJGdW1dJxyo7T0nlVVd0/yeYAVXVVkh3XKUJJkiRJ0gZhnGs6Pw5dsjkweuzJ/YUkSZIkSVooRrZ0JrkPsBuwRZKnDizaHNis78AkSZIkSeu/mbrX3ht4MrAlsM/A/KuBF/YZlCRJkiRpYRiZdFbVp4BPJXl4VX1zDmOSJEmSJC0QaxxIyIRT0rpydFVJkqQN1zgDCUmSJEmStFZGJp1JXtb+PmLuwpEkSZIkLSQztXQ+t/1911wEIkmarGTtHpIkSTOZ6ZrOc5NcACxO8r2B+QGqqu7Xa2SSJEmSpPXeTKPXHpjk7sApwL5zF5IkSZIkaaGYcfTaqvolcP8kmwC7tNnnVdUNvUcmrYccpVWSJEm6uTXeMiXJo4EPARfQda1dkuSgqvpqz7FJkiRJktZza0w6gbcDe1bVeQBJdgE+Cjy4z8AkSZIkSeu/ce7TufFUwglQVT8GNu4vJEmSJEnSQjFOS+fyJMcAx7fpZwBn9heSJEmSJGmhGCfpfAnwN8BL6a7p/Crw3j6DkiRJkiQtDGtMOqvqt3TXdb69/3AkSZIkSQvJOC2d0gbB251IkiRJs2+cgYQkSZIkSVorJp2SJEmSpN6ssXttuy/nq4DtB8tX1WN7jEuSJEmStACMc03nx4CjgPcBN/YbjiRJkiRpIRkn6VxdVf/eeySSJEmSpAVnnGs6P53kr5PcI8mdpx7jbDzJXknOS7IiyaFDlm+a5MS2/PQkO0xbvl2Sa5L8/VhHI0mSJEmaV8Zp6Tyo/X3VwLwC7jnTSkk2At4DPB5YCZyRZFlV/XCg2POBy6tqpyQHAEcATx9YfiTw+TFilCRJkiTNQ2tMOqtqx7Xc9u7Aiqo6HyDJCcB+wGDSuR9wWHt+MvDuJKmqSvIU4Hzg2rXcvyRJkiRpwtbYvTbJxklemuTk9jgkycZjbHsb4KKB6ZVt3tAyVbUauBLYKsntgVcDbxjnICRJkiRJ89M413T+O/Bg4L3t8eA2b00yZF6NWeYNwJFVdc2MO0gOTrI8yfJVq1aNEZIkSZIkaS6Nc03nQ6rq/gPTpyU5e4z1VgJLBqa3BS4eUWZlkkXAFsBlwEOB/ZO8BdgS+H2S66vq3YMrV9XRwNEAS5cunZ7QSpIkSZImbJyk88Yk96qqnwAkuSfj3a/zDGDnJDsCPwcOAP5qWplldAMVfRPYHzitqgr406kCSQ4DrpmecEqSJEmS5r9xks5XAV9Kcj5dd9jtgeeuaaWqWp3kEOAUYCPg2Ko6J8nhwPKqWgYcAxyfZAVdC+cBa3kckiRJkqR5KF3D4hoKJZsC96ZLOn9UVb/tO7Bba+nSpbV8+fJJh6H1WIZdYTyGwbfQbGzDWPqNZba2YyyjtyNJkha+JGdW1dJxyo5s6Uzy2Ko6LclTpy26VxKq6hPrFKUkSZIkacGbqXvto4HTgH2GLCvApFOSJEmSNKORSWdVvb49Pbyqfjq4rA0OJEmSJEnSjMa5T+fHh8w7ebYDkSRJkiQtPDNd03kfYDdgi2nXdW4ObNZ3YJIkSZKk9d9M13TeG3gysCU3v67zauCFfQYlSZIkSVoYZrqm81PAp5I8vKq+OYcxSZIkSZIWiHGu6Xxxki2nJpLcKcmxPcYkSZIkSVogxkk671dVV0xNVNXlwAP7C0mSJEmStFCMk3TeJsmdpiaS3JmZrwWVJEmSJAkYL3n8N+AbSaZuk/I04I39hSRJkiRJWijWmHRW1YeSnAk8Bgjw1Kr6Ye+RSZIkSZLWe2N1k62qc5Ksot2fM8l2VfWzXiOTJEmSJK331nhNZ5J9k/wf8FPgK8AFwOd7jkuSJEmStACMM5DQPwMPA35cVTsCewD/22tUkiRJkqQFYZyk84aqupRuFNvbVNWXgAf0HJckSZIkaQEY55rOK5LcAfgq8JEkvwZW9xuWJEmSJGkhGKelcz/gOuAVwH8DPwH26TMoSZIkSdLCMGNLZ5KNgE9V1eOA3wPHzUlUkqT1UrJ261XNbhySJGn+mLGls6puBK5LssUcxSNJkiRJWkDGuabzeuD7Sb4AXDs1s6pe2ltUkiRJkqQFYZyk87PtIUmSJEnSrTIy6UzyxaraA9i1ql49hzFJkrTOvL5UkqT5YaaWznskeTSwb5ITgJv9+66qs3omb+GoAAAVRElEQVSNTJIkSZK03psp6XwdcCiwLfD2acsKeGxfQUmSJEmSFoaRo9dW1clVtTfwlqp6zLTHWAlnkr2SnJdkRZJDhyzfNMmJbfnpSXZo83dP8t32ODvJn6/l8UmSJEmSJmjGW6YAVNU/r82G2z0+3wPsDewKHJhk12nFng9cXlU7AUcCR7T5PwCWVtUDgL2A/0gyzqBHkiRJkqR5ZI1J5zrYHVhRVedX1e+AE4D9ppXZDziuPT8Z2CNJquq6qlrd5m9G151XkiRJkrSe6TPp3Aa4aGB6ZZs3tExLMq8EtgJI8tAk5wDfB148kITeJMnBSZYnWb5q1aoeDkHrg2TtHpIkSZL612fSOexr/fQWy5Flqur0qtoNeAjwmiSb3aJg1dFVtbSqli5evHidA5YkSZIkza61SjqTfGaMYiuBJQPT2wIXjyrTrtncArhssEBVnQtcC/zR2sQqSZIkSZqctW3pfOEYZc4Adk6yY5JNgAOAZdPKLAMOas/3B06rqmrrLAJIsj1wb+CCtYxVkiRJkjQha0w6k7xsyOy/XNN67RrMQ4BTgHOBk6rqnCSHJ9m3FTsG2CrJCuCVdPcFBXgkcHaS7wKfBP66qi5Z49FIkiRJkuaVVM08MGySs6rqQdPmfaeqHthrZLfS0qVLa/ny5ZMOQxOwtoMCTa/6s7EdY5n/sczWdoyl31hmw3yKRZKkhSbJmVW1dJyyI+99meRA4K+AHZMMdovdHLh03UKUJEmSJG0IRiadwDeAXwB3Af5tYP7VwPf6DEqSJEmStDCMTDqr6kLgwiSPA35TVb9PsgtwH7p7Z0qSJEmSNKNxRq/9KrBZkm2ALwLPBT7YZ1CSJEmSpIVhnKQzVXUd8FTgXVX158Cu/YYlSdqQJWv3kCRJ889YSWeShwPPAD7b5s10LagkSZIkScB4SefLgdcAn2z32bwn8KV+w5IkSZIkLQRrbLGsqq8AXxmYPh94aZ9BSZIkSZIWhpnu0/mOqnp5kk8Dt7hVdlXt22tkkiRJkqT13kwtnce3v2+bi0AkSZIkSQvPTPfpPLP9/UqSxe35qrkKTJIkSZK0/hs5kFA6hyW5BPgR8OMkq5K8bu7CkyRp/ectYCRJG7KZRq99OfAI4CFVtVVV3Ql4KPCIJK+Yk+gkSZIkSeu1mZLOZwMHVtVPp2a0kWuf2ZZJkiRJkjSjmZLOjavqkukz23WdG/cXkiRJkiRpoZgp6fzdWi6TJEmSJAmY+ZYp909y1ZD5ATbrKR5JkiRJ0gIy0y1TNprLQCRJkiRJC89MLZ2SJG3w1vbWJVWzG4ckSeurma7plCRJkiRpnZh0SpIkSZJ6Y9IpSZIkSeqNSackSZIkqTcmnZIkSZKk3vSadCbZK8l5SVYkOXTI8k2TnNiWn55khzb/8UnOTPL99vexfcYpSZIkSepHb0lnko2A9wB7A7sCBybZdVqx5wOXV9VOwJHAEW3+JcA+VfXHwEHA8X3FKUmSJEnqT58tnbsDK6rq/Kr6HXACsN+0MvsBx7XnJwN7JElVfaeqLm7zzwE2S7Jpj7FKkiRJknrQZ9K5DXDRwPTKNm9omapaDVwJbDWtzF8A36mq3/YUpyRJkiSpJ4t63HaGzKtbUybJbnRdbvccuoPkYOBggO22227topQkSZIk9abPls6VwJKB6W2Bi0eVSbII2AK4rE1vC3wSeHZV/WTYDqrq6KpaWlVLFy9ePMvhS5IkSZLWVZ9J5xnAzkl2TLIJcACwbFqZZXQDBQHsD5xWVZVkS+CzwGuq6n97jFGSJEmS1KPeks52jeYhwCnAucBJVXVOksOT7NuKHQNslWQF8Epg6rYqhwA7Af+U5Lvtcde+YpUkSZIk9SNV0y+zXD8tXbq0li9fPukwNAEZdmXwGKZX/dnYjrHM/1hmazvGYiyTiEWSpPkiyZlVtXScsn12r5UkSZIkbeBMOiVJkiRJvTHplCRJkiT1xqRTkiRJktQbk05JkiRJUm8WTToAbdgc0VGSJEla2GzplCRJkiT1xqRTkiRJktQbu9dKkrSe8JIESdL6yJZOSZIkSVJvTDolSZIkSb0x6ZQkSZIk9cakU5IkSZLUG5NOSZIkSVJvTDolSZIkSb3xlimSJGkivAWMJG0YbOmUJEmSJPXGpFOSJEmS1BuTTkmSJElSb0w6JUmSJEm9MemUJEmSJPXGpFOSJEmS1BuTTkmSJElSb0w6JUmSJEm9MemUJEmSJPWm16QzyV5JzkuyIsmhQ5ZvmuTEtvz0JDu0+Vsl+VKSa5K8u88YJUmSJEn96S3pTLIR8B5gb2BX4MAku04r9nzg8qraCTgSOKLNvx74J+Dv+4pPkiRJktS/Pls6dwdWVNX5VfU74ARgv2ll9gOOa89PBvZIkqq6tqq+Tpd8SpIkSZLWU30mndsAFw1Mr2zzhpapqtXAlcBW4+4gycFJlidZvmrVqnUMV5IkSZI02/pMOjNkXq1FmZGq6uiqWlpVSxcvXnyrgpMkSZIk9a/PpHMlsGRgelvg4lFlkiwCtgAu6zEmSZIkSdIc6jPpPAPYOcmOSTYBDgCWTSuzDDioPd8fOK2qxm7plCRJkiTNb4v62nBVrU5yCHAKsBFwbFWdk+RwYHlVLQOOAY5PsoKuhfOAqfWTXABsDmyS5CnAnlX1w77ilSRpQ5BhF7aMwZ+EJUlrq7ekE6CqPgd8btq81w08vx542oh1d+gzNkmSJElS//rsXitJkiRJ2sD12tKp+cmuVZKkdTWf/pfMp1gkSbdkS6ckSZIkqTcmnZIkSZKk3ti9VmvFrkySJEmSxmFLpyRJkiSpNyadkiRJkqTemHRKkiRJknrjNZ2SJEk4XoEk9cWWTkmSJElSb0w6JUmSJEm9MemUJEmSJPXGpFOSJEmS1BuTTkmSJElSbxy9VpIkSSM5qq+kdWVLpyRJkiSpNyadkiRJkqTemHRKkiRJknrjNZ3rEa+pkCRpwzAb//P93jDcfDov8ymW+cTzsvDY0ilJkiRJ6o1JpyRJkiSpN3avlSRJmiV2C+yX53c4z4vmO1s6JUmSJEm9MemUJEmSJPWm16QzyV5JzkuyIsmhQ5ZvmuTEtvz0JDsMLHtNm39ekif0GedcSNbuIUmStL5biN+DFuIxzQbPy3Ab+nnpLelMshHwHmBvYFfgwCS7Tiv2fODyqtoJOBI4oq27K3AAsBuwF/Detj1JkiRJ0nqkz5bO3YEVVXV+Vf0OOAHYb1qZ/YDj2vOTgT2SpM0/oap+W1U/BVa07UmSJEmS1iN9Jp3bABcNTK9s84aWqarVwJXAVmOuK0mSJEma5/q8ZcqwXsjTB2YeVWacdUlyMHBwm7wmyXm3KsL54y7AJcMWzEZf7lu5DWMxFmMZ4VZsx1iMxVgWViywAX3eGYux9Gi9OC/zyTy/rnP7cQv2mXSuBJYMTG8LXDyizMoki4AtgMvGXJeqOho4ehZjnogky6tq6aTjAGMZxViGM5bhjGU4YxnOWIabT7HA/IrHWIYzluGMZbj5FMuGoM/utWcAOyfZMckmdAMDLZtWZhlwUHu+P3BaVVWbf0Ab3XZHYGfg2z3GKkmSJEnqQW8tnVW1OskhwCnARsCxVXVOksOB5VW1DDgGOD7JCroWzgPauuckOQn4IbAa+JuqurGvWCVJkiRJ/eizey1V9Tngc9PmvW7g+fXA00as+0bgjX3GN4/Mpy7CxjKcsQxnLMMZy3DGMpyxDDefYoH5FY+xDGcswxnLcPMplgUvXW9WSZIkSZJmX5/XdEqSJEmSNnAmnXMoyWZJvp3k7CTnJHlDm79HkrOSfDfJ15PsNIcxbZTkO0k+06Z3THJ6kv9LcmIbBGpSsXwkyXlJfpDk2CQbTyqWgfnvSnLNXMUxLJZ03pjkx0nOTfLSCcYyybp7QZLvt30vb/PunOQLrf5+IcmdJhjLW5P8KMn3knwyyZaTimVg2d8nqSR3mWQsSf62vbfPSfKWScWS5AFJvjU1L8nucxTLlklObvXj3CQPn2DdHRbLpOruLWIZWDbXdXdoLBOqu8Neozmvu0nu3fY39bgqycsnUXdniGXO6+6oWAaWz1ndnSmWua67M7xGk/rcfUU79h8k+Wi67+Q7ZkLfeTdIVeVjjh509x+9Q3u+MXA68DDgx8B92/y/Bj44hzG9EvhP4DNt+iTggPb8KOAlE4zlie2cBfjoJGNp85YCxwPXzHG9mX5engt8CLhNm77rBGOZZN29ALjLtHlvAQ5tzw8FjphgLHsCi9rzIyYZS5u/hG5gtwuHLZ/D8/IY4H+ATdv0nNTfEbGcCuzdnj8R+PIcxXIc8IL2fBNgywnW3WGxTKru3iKW9nwSdXfYeZlU3R0Wy0Tq7kBMGwG/pLtH30Tq7ohYJlJ3h8XSpue87o44LxOpuyNimfO6C2wD/BS4bZs+CXgOE/zOuyE+bOmcQ9WZaiXbuD2qPTZv87dgyD1J+5BkW+BJwPvbdIDHAie3IscBT5lELNANRNXOWdHdMmfbScWSZCPgrcA/zEUMM8UCvAQ4vKp+D1BVv55gLBOpuzPYj67ewhzW32Gq6tSqWt0mv8Uc1d8ZHElXfyd9If9LgDdX1W9h7urvCHNef5NsDjyKbvR2qup3VXUFE6i7o2KZRN2d4bzAHNfdGWKZ87o7QyyT/uzdA/hJVV3I5D93b4plHnzuDp4XmOzn7mAsk/7cHYxlUnV3EXDbJIuA2wG/YELfeTdUJp1zLF33xO8Cvwa+UFWnAy8APpdkJfAs4M1zFM476D4Mf9+mtwKuGPjAXkn369AkYrlJum61zwL+e4KxHAIsq6pfzFEMM8VyL+DprVvK55PsPMFYJlV3ofvHdWqSM5Mc3Obdbeo1an/vOsFYBj0P+PykYkmyL/Dzqjp7jmIYGQuwC/CnrUvTV5I8ZIKxvBx4a5KLgLcBr5mDOO4JrAI+kK6r+vuT3J7J1N1RsQyaq7o7NJYJ1d1R52USdXdULJOou4MOoOuBBJP73B0Wy6C5/Ny9RSwT/Ny9RSxM7nN3WCxzXner6udtXz+jSzavBM5kct95N0gmnXOsqm6sqgfQ/fq2e5I/Al4BPLGqtgU+ALy97ziSPBn4dVWdOTh7WMgTimXQe4GvVtXXJhFLkq3pbu3zrr73v6ZYmk2B66tqKfA+4NgJxjLndXfAI6rqQcDewN8kedQc7nvsWJK8lu5+wx+ZYCyvBV4382pzFssi4E50lxa8Cjip9bKYRCwvAV5RVUvo6vIxcxDHIuBBwL9X1QOBa+m6JE7CjLHMcd0dFsthTKbujjovk6i7o2KZRN0FoF33ti/wsbna562NZQKfuzeLJcntmNzn7rDzMqnP3WGxzHndTXed8X7AjsDWwO3p/hdMN+meQAuaSeeEtO4xX6ar9PdvLZ4AJwJ/MgchPALYN8kFwAl0XQzeAWzZuh5AlxjPRbeHW8SS5MMASV4PLKa7lnAuDDsv5wA7ASva/NslWTGJWNp5WQl8vJX5JHC/CcXyWSZTdwGoqovb31/TnYfdgV8luQdA+zsnXYhGxEKSg4AnA89o3cQnEcuj6f7Rnt1ev22Bs5LcfQKx7E5Xfz/Res5/m67lvPcBNkbEchDwiVbkY21e31YCKwfeNyfTJRWTqLujYplE3R0VyyTq7qhYJlF3R8Uyibo7ZW/grKr6VZueyOfuiFgm8rk7JJZ7MaHP3SGxwIQ+d0fEMom6+zjgp1W1qqpuaPv/EybznXeDZdI5h5IsThtJLclt6d4E5wJbJNmlFXt8m9erqnpNVW1bVTvQdXs4raqeAXwJ2L8VOwj41IRieWaSFwBPAA6cun5xQrHcqaruXlU7tPnXVVXvo7SOOi/Af9Elw9AlFD+eRCx0vxrOed0FaN3u7jj1nG7wiB8Ay+jqLcxR/R0VS5K9gFcD+1bVdX3HMUMsZ1TVXQfq70rgQVX1ywnE8gMG6m+rO5sAl0wolovp3kO0mP6vzzgA2nm/KMm926w9gB8ygbo7KpZJ1N0RsZw1ibo7w2s053V3hljmvO4OOJCbd2ed87o7KpZJ1N1hsVTV9ydRd4fF0sx53Z0hlknU3Z8BD0tyu9bCO/U+mvPvvBuyRWsuoll0D+C4dIPS3AY4qao+k+SFwMeT/B64nO46hEl5NXBCkn8BvsMcdtkZ4ii6Ed++2XqBfKKqDp9gPPPFm4GPJHkFcA3ddZVzrqpWT7Du3g34ZKsXi4D/rKr/TnIGXbeh59P9k3naBGNZQdcV+gtt2beq6sWTiKXnfd6qWFpXq2OT/AD4HXDQHLRGjIrlGuCd7Zfu64Fh1+P24W/p3sObAOfTjUh9G+a+7o6K5Qzmvu6OimVShsVyLXNfd0fF8ikmUHdbt9HHAy8amP1mJlB3R8TybiZQd0fEMhEjYjmWCdTdEbG8kDmuu1V1epKTgbPoul1/Bzga+Czz5zvvgpe57XkgSZIkSdqQ2L1WkiRJktQbk05JkiRJUm9MOiVJkiRJvTHplCRJkiT1xqRTkiRJktQbk05JktZRktcmOSfJ95J8N8lDJx2TJEnzhffplCRpHSR5OPBkupu//zbJXehuvr6221tUVatnLUBJkibMlk5JktbNPYBLquq3AFV1SVVdnOQhSb6R5Owk305yxySbJflAku8n+U6SxwAkeU6SjyX5NHBqm/eqJGe01tM3tHm3T/LZts0fJHn6pA5akqRx2dIpSdK6ORV4XZIfA/8DnAh8s/19elWdkWRz4DfAywCq6o+T3Ac4NckubTsPB+5XVZcl2RPYGdgdCLAsyaOAxcDFVfUkgCRbzNlRSpK0lmzplCRpHVTVNcCDgYOBVXTJ5ouAX1TVGa3MVa3L7COB49u8HwEXAlNJ5xeq6rL2fM/2+A5wFnAfuiT0+8DjkhyR5E+r6so5OERJktaJLZ2SJK2jqroR+DLw5STfB/4GqCFFM8Nmrp1W7k1V9R+32EDyYOCJwJuSnFpVh6914JIkzQFbOiVJWgdJ7p1k54FZDwDOBbZO8pBW5o5JFgFfBZ7R5u0CbAecN2SzpwDPS3KHVnabJHdNsjVwXVV9GHgb8KC+jkuSpNliS6ckSevmDsC7kmwJrAZW0HW1/UCbf1u66zkfB7wXOKq1hq4GntNGvL3ZBqvq1CT3Bb7Zll0DPBPYCXhrkt8DNwAvmYPjkyRpnaRqWO8fSZIkSZLWnd1rJUmSJEm9MemUJEmSJPXGpFOSJEmS1BuTTkmSJElSb0w6JUmSJEm9MemUJEmSJPXGpFOSJEmS1BuTTkmSJElSb/5/gPRpMQfULmMAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "## runtime of this part could be take 5 to 10 mins\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "#Using written functions to create distributions.\n",
    "\n",
    "distribution_unnormalized = generate_null_distribution(human_gene, fruitfly_gene, pam_50_matrix, 1000)\n",
    "distribution_normalized = normalize_distribution(distribution_unnormalized)\n",
    "distribution_normalized\n",
    "\n",
    "#Craeating the plot\n",
    "\n",
    "keys_normalized = distribution_normalized.keys()\n",
    "values_normalized = [round(val,3) for val in distribution_normalized.values()]\n",
    "\n",
    "fig = plt.figure()\n",
    "ax = fig.add_axes([0,0,2,1])\n",
    "ax.bar(keys_normalized, values_normalized, color='blue')\n",
    "ax.set_xticks(np.arange(keys_normalized[0], keys_normalized[-1], 2))\n",
    "ax.set_ylabel('Dist. fraction of total trials')\n",
    "ax.set_xlabel('Scores')\n",
    "ax.set_title('Scores by normalized dist. fractions of the experiment')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Distribution resulted in a bell shaped graph as expected, and the scores concentrated mostly around 40 - 60 range. After distribution calculated, simple statistical analysis would be helpful to determine the likelihood of the score 875, found between local alignments of human and fruit fly eyeless genes. In order to do that, mean, variance and z-score will be calculated with the functions written below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(51.618000000000016, 6.6765317343662804, 123.32480886173056)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def calculate_mean_std_zscore(normalized_dist,score):\n",
    "    \"\"\"\n",
    "    Gets distribution and calculated score and returns mean,\n",
    "    standard deviation and zscore of the given dist.\n",
    "    \"\"\"\n",
    "    mean = 0\n",
    "    total = 0\n",
    "    \n",
    "    for key in normalized_dist.keys():\n",
    "        mean += normalized_dist[key]*key\n",
    "    \n",
    "    for key_score in  normalized_dist.keys():\n",
    "        total += ((key_score-mean)**2)*distribution_normalized[key_score]\n",
    "        \n",
    "    std = (total)**(1/2.0)\n",
    "    \n",
    "    z_score = (score - mean)/float(std)\n",
    "    \n",
    "    return (mean,std,z_score)\n",
    "    \n",
    "calculate_mean_std_zscore(distribution_normalized,alignment_loc[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Implementation of the function resulted with 51.2 mean, 6.67 standard deviation and 123.4 z-score. So the score, that is calculated between the local alignment of human and fruit fly genes, lies in a distance of 123 standard deviation ahead to the mean of the simulated event. The probability of such a situation is around 10^(-3288) as can be seen blow:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![title](img/1618769562195.png) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we think probability of winning a lottery is about one in a trillion 10^(-12). This probability above gives an estimate about the likelihood of such a similarity to occur randomly between the local alignments of human and fruit fly eyeless gene."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
